### PR TITLE
CI: Add arm64 in travis-ci

### LINKED ISSUE #11170: Use manylinux2014 to get aarch64/ppc64le support
Manylinux2014 image got released some time ago. One of things it brings is support for !x86 architectures.

Scipy is using Travis CI so two things can be joined and used to generate wheel files for aarch64 and ppc64le architectures.

Amount of software installed to be able to run 'pip install scipy' would be cut. Also install times would be nicer.

### PR DESCRIPTION
**Pull request by odidev**
#### Reference issue

see #11170 

#### What does this implement/fix?

Added the arm64 job for Python v3.7 in Travis-ci.

**Additional information**

As per the discussion with @rgommers, included arm64 job for only ```TESTMODE=fast```.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -36,6 +36,17 @@ matrix:
...
+    - python: 3.7
+      if: type = pull_request
+      arch: arm64
+      dist: bionic
+      env:
+        - TESTMODE=fast
+        - COVERAGE=
+        # Need a test with most recent Python version where we build from an
+        # sdist (uses pip build isolation), to check that the version we
+        # specify in pyproject.toml (will be older than --pre --upgrade) works
```
Comment (rgommers) [inline .travis.yml:48]: This comment can be removed, it's copied from the other `USE_SDIST=1` test, no need to have it twice.

I am +1 on keeping `USE_SDIST=``.

File: .travis.yml [hunk 2]
```diff
@@ -116,7 +128,7 @@ cache:
 before_install:
   - echo $TRAVIS_OS_NAME
   - |
-    if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
+    if [[ "$TRAVIS_OS_NAME" == "linux" && "$TRAVIS_CPU_ARCH" == "amd64" ]]; then
```
Comment (rgommers) [inline .travis.yml:131]: Doesn't `ccache` work on ARM64? Should speed things up ....

If it really doesn't, can you add a comment here about what the problem is?

File: .travis.yml [hunk 3]
```diff
@@ -165,18 +183,33 @@ before_install:
...
+  - if [[ "${TRAVIS_CPU_ARCH}" == "arm64" ]]; then
+      wget -q "https://github.com/conda-forge/miniforge/releases/download/4.8.2-1/Miniforge3-4.8.2-1-Linux-aarch64.sh"  -O miniconda.sh;
+      chmod +x miniconda.sh;
+      ./miniconda.sh -b -p $HOME/miniconda3;
+      export PATH=$HOME/miniconda3/bin:$PATH;
+      conda config --set always_yes yes --set auto_update_conda false;
+      conda install pip conda;
+      conda update -n base conda;
+      conda info -a;
+      conda install pytest pytest-xdist scipy mpmath Cython pybind11 coverage;
```
Comment (rgommers) [inline .travis.yml:202]: no need for `coverage` here I believe

File: .travis.yml [hunk 4]
```diff
@@ -165,18 +183,33 @@ before_install:
...
+    fi
 script:
   - python -c 'import numpy as np; print("relaxed strides checking:", np.ones((10,1),order="C").flags.f_contiguous)'
   # Make sure that relaxed strides checking is actually in effect; otherwise fail loudly
   - if [ "$NPY_RELAXED_STRIDES_CHECKING" == "1" ]; then python -c'import numpy as np; assert np.ones((10,1),order="C").flags.f_contiguous'; fi
   # Test that mpmath actually uses gmpy2
-  - if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then python -c 'import mpmath.libmp; assert mpmath.libmp.BACKEND == "gmpy"'; fi
   - |
-    if [ "${USE_WHEEL}" == "1" ]; then
+    if [[ "$TRAVIS_OS_NAME" == "linux" && "$TRAVIS_CPU_ARCH" != "arm64" ]]; then python -c 'import mpmath.libmp; assert mpmath.libmp.BACKEND == "gmpy"'; fi
```
Comment (rgommers) [inline .travis.yml:210]: Could you split up this line? It runs off the screen now

File: .travis.yml [hunk 5]
```diff
@@ -199,8 +232,14 @@ script:
         cd ..
         USE_WHEEL_BUILD="--no-build"
     fi
-  - export SCIPY_AVAILABLE_MEM=3G
-  - python -u runtests.py -g -m $TESTMODE $COVERAGE $USE_WHEEL_BUILD -- -rfEX -n 3 2>&1 | tee runtests.log
+  - |
+    if [[ "$TRAVIS_CPU_ARCH" = "arm64" ]]; then
```
Comment (rgommers) [inline .travis.yml:236]: should be `==` rather than `=` I think?

File: .travis.yml [hunk 6]
```diff
@@ -165,18 +183,33 @@ before_install:
...
   - python -c 'import numpy as np; print("relaxed strides checking:", np.ones((10,1),order="C").flags.f_contiguous)'
   # Make sure that relaxed strides checking is actually in effect; otherwise fail loudly
   - if [ "$NPY_RELAXED_STRIDES_CHECKING" == "1" ]; then python -c'import numpy as np; assert np.ones((10,1),order="C").flags.f_contiguous'; fi
   # Test that mpmath actually uses gmpy2
-  - if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then python -c 'import mpmath.libmp; assert mpmath.libmp.BACKEND == "gmpy"'; fi
   - |
-    if [ "${USE_WHEEL}" == "1" ]; then
+    if [[ "$TRAVIS_OS_NAME" == "linux" && "$TRAVIS_CPU_ARCH" != "arm64" ]]; then python -c 'import mpmath.libmp; assert mpmath.libmp.BACKEND == "gmpy"'; fi
+  - |
+    if [ "${USE_WHEEL}" == "1" -a "${TRAVIS_CPU_ARCH}" != "arm64" ]; then
```
Comment (rgommers) [inline .travis.yml:212]: This change is unneeded, since we don't have `USE_WHEEL=1` + ARM64.

File: .travis.yml [hunk 7]
```diff
@@ -185,7 +218,7 @@ script:
         python tools/suppress_output.py pip wheel --no-build-isolation .
         pip install --no-cache-dir scipy*.whl
         USE_WHEEL_BUILD="--no-build"
-    elif [ "${USE_SDIST}" == "1" ]; then
+    elif [ "${USE_SDIST}" == "1" -a "${TRAVIS_CPU_ARCH}" != "arm64"  ]; then
```
Comment (rgommers) [inline .travis.yml:221]: Here the logic gets messed up. In the ARM64 build job definition `USE_SDIST=1` is set, however here it's being bypassed. There are two options here:
1. Remove `USE_SDIST=1` and undo this change, or
2. just undo this change 

I'd prefer (2).

File: .travis.yml [hunk 8]
```diff
@@ -199,8 +232,14 @@ script:
         cd ..
         USE_WHEEL_BUILD="--no-build"
     fi
-  - export SCIPY_AVAILABLE_MEM=3G
-  - python -u runtests.py -g -m $TESTMODE $COVERAGE $USE_WHEEL_BUILD -- -rfEX -n 3 2>&1 | tee runtests.log
+  - |
+    if [[ "$TRAVIS_CPU_ARCH" = "arm64" ]]; then
+      python -u runtests.py --parallel=4 | tee runtests.log
```
Comment (rgommers) [inline .travis.yml:237]: Some of the options of the original command should be kept. Changing only what's necessary would be useful. Which is very little I think. The only difference needed seems to be adding `--parallel=4` and removing `n 3` (`-n` means `--numprocesses` to `pytest`).

`-g` does nothing, the other env vars are useful. Can't find the exact meaning for `2>&1`; `-rfEX` needs to be kept I believe.

**Review by rgommers — CHANGES_REQUESTED** [review]
Thanks @odidev. Great to see this passes. I think the logic needs cleaning up a little to keep this maintainabe and consistent between platforms.

File: .travis.yml (hunk 8 above)
Comment (andyfaff) [reply .travis.yml:237]: 2>&1 redirects stderr to stdout

**Comment by andyfaff** [comment]
It's great to have an arm64 build. My overall impression was that there are a lot of `if` logic statements. I wonder if there's a cleaner way of structuring the file.

File: .travis.yml (hunk 8 above)
Comment (ilayn) [reply .travis.yml:237]: `rfEX` is for pytest [output report](https://docs.pytest.org/en/latest/usage.html#detailed-summary-report)

> The -r options accepts a number of characters after it, with a used above meaning “all except passes”.
> 
> Here is the full list of available characters that can be used:
> 
>         f - failed
>         E - error
>         s - skipped
>         x - xfailed
>         X - xpassed
>         p - passed
>         P - passed with output
> 
>

**Comment by odidev** [comment]
@rgommers @andyfaff I have incorporated the changes you suggested in the recent commit, Please have a look and do revert if I have missed something or any other changes are required to be done.

**Comment by snnn** [comment]
I'm wondering how did you get numpy work.  I was stuck at a compiler error. 
See:
numpy/numpy#13622

**Comment by rgommers** [comment]
> I'm wondering how did you get numpy work. I was stuck at a compiler error.
> See:
> [numpy/numpy#13622](https://github.com/numpy/numpy/issues/13622)

@snn there's lots of possible workaround in that issue: use gcc 9, compile with `-O0`, apply the last patch on that issue, or simply install numpy from `conda-forge.

File: .travis.yml [hunk 9]
```diff
@@ -36,6 +36,15 @@ matrix:
...
           grep -rlIP '[^\x00-\x7F]' scipy | grep '\.pyx\?' | sort > unicode.out; grep -rlI '# -\*- coding: \(utf-8\|latin-1\) -\*-' scipy | grep '\.pyx\?' | sort > coding.out; comm -23 unicode.out coding.out > test_code.out; cat test_code.out;  test \! -s test_code.out
+    - python: 3.7
+      if: type = pull_request
+      arch: arm64
+      dist: bionic
+      env:
+        - TESTMODE=fast
+        - COVERAGE=
+        - USE_SDIST=1
+        - PIP_NO_CACHE_DIR=off
```
Comment (rgommers) [inline .travis.yml:47]: This env var exists but is basically undocumented. If I read it correctly, this _disables turning off_, so turns the cache on - which is the default I think?

File: .travis.yml [hunk 10]
```diff
@@ -165,16 +181,33 @@ before_install:
...
+      conda install pytest pytest-xdist scipy mpmath Cython pybind11;
+    fi
 script:
   - python -c 'import numpy as np; print("relaxed strides checking:", np.ones((10,1),order="C").flags.f_contiguous)'
   # Make sure that relaxed strides checking is actually in effect; otherwise fail loudly
   - if [ "$NPY_RELAXED_STRIDES_CHECKING" == "1" ]; then python -c'import numpy as np; assert np.ones((10,1),order="C").flags.f_contiguous'; fi
   # Test that mpmath actually uses gmpy2
-  - if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then python -c 'import mpmath.libmp; assert mpmath.libmp.BACKEND == "gmpy"'; fi
+  - |
+    if [[ "$TRAVIS_OS_NAME" == "linux" && "$TRAVIS_CPU_ARCH" != "arm64" ]]; then
```
Comment (rgommers) [inline .travis.yml:208]: Is this not working on arm64?

File: .travis.yml [hunk 11]
```diff
@@ -136,25 +146,31 @@ before_install:
...
+  - |
+    if [ "${TRAVIS_CPU_ARCH}" != "arm64" ]; then
+        travis_retry pip install --upgrade pip setuptools wheel
+        travis_retry pip install cython
+    fi
+  - |
+    if [[ "$NUMPYSPEC" && "$TRAVIS_CPU_ARCH" != "arm64" ]]; then travis_retry pip install $NUMPYSPEC; fi
+  - |
+    if [ "${TRAVIS_CPU_ARCH}" != "arm64" ]; then
+        travis_retry pip install --upgrade pytest pytest-xdist mpmath argparse Pillow codecov
```
Comment (rgommers) [inline .travis.yml:158]: None of these packages have a dependency on `numpy`, so this can be moved into the `f [ "${TRAVIS_CPU_ARCH}" != "arm64" ];` above to make things less verbose

File: .travis.yml [hunk 12]
```diff
@@ -136,25 +146,31 @@ before_install:
...
+    fi
   - |
     if [ -n "${USE_DEBUG}" ]; then
         # see gh-10676; need to pin pytest version with debug
         # Python 3.6
         travis_retry pip install pytest==5.0.1
     fi
-  - travis_retry pip install gmpy2  # speeds up mpmath (scipy.special tests)
-  - travis_retry pip install pybind11
+  - if [[ "${TRAVIS_CPU_ARCH}" != "arm64" ]]; then travis_retry pip install gmpy2 pybind11; fi
```
Comment (rgommers) [inline .travis.yml:166]: Same here, let's move into that same block

File: .travis.yml [hunk 13]
```diff
@@ -136,25 +146,31 @@ before_install:
...
   - |
     if [ -z "${USE_DEBUG}" -a "${TRAVIS_CPU_ARCH}" == "amd64" -a "${TRAVIS_PYTHON_VERSION}" != "3.8" ]; then
         travis_retry pip install numba==0.47
         travis_retry pip install --only-binary=:all: sparse
     fi
   - |
-    if [ "${TESTMODE}" == "full" ]; then
+    if [ "${TESTMODE}" == "full" -a "${TRAVIS_CPU_ARCH}" != "arm64" ]; then
```
Comment (rgommers) [inline .travis.yml:173]: This change can be undone, it's not needed because there's no arm64 test with "full".

**Comment by rgommers** [comment]
Looks like we're almost there. I'd just like to reduce the if-statements a little. @odidev if you could look at the remaining questions/comments, then I think we can merge this soon.

File: .travis.yml (hunk 10 above)
Comment (odidev) [reply .travis.yml:200]: yes it is working on arm64 hence I have removed the arm64 related check.

**Comment by odidev** [comment]
@rgommers I have incorporated the changes you suggested, Please have a look. If something else is required to be done please do revert.

File: .travis.yml (hunk 9 above)
Comment (odidev) [reply .travis.yml:47]: @rgommers I have added this var because on arm64 as I was getting the below issue in order to resolve the same I have added this: 

```
 WARNING: Building wheel for scipy failed: [Errno 13] Permission denied: '/home/travis/.cache/pip/wheels/37'
Failed to build scipy
Cleaning up...
  Removing source in /tmp/pip-req-build-jfbnl8mm
Removed build tracker: '/tmp/pip-req-tracker-95zki2iq'
ERROR: Could not build wheels for scipy which use PEP 517 and cannot be installed directly
Exception information:
Traceback (most recent call last):
  File "/home/travis/miniconda3/lib/python3.7/site-packages/pip/_internal/cli/base_command.py", line 186, in _main
    status = self.run(options, args)
  File "/home/travis/miniconda3/lib/python3.7/site-packages/pip/_internal/commands/install.py", line 375, in run
    ", ".join(r.name for r in pep517_build_failures)))
pip._internal.exceptions.InstallationError: Could not build wheels for scipy which use PEP 517 and cannot be installed directly
```

Comment (rgommers) [reply .travis.yml:47]: Interesting, looks like a TravisCI bug. Thanks for explaining @odidev

**Comment by rgommers** [comment]
Grr, timed out again: `The job exceeded the maximum time limit for jobs, and has been terminated.` At `82%`. Pushed some fixes and re-try.

File: .travis.yml (hunk 10 above)
Comment (rgommers) [reply .travis.yml]: it was still there, so I reverted this change

File: .travis.yml [hunk 14]
```diff
@@ -136,25 +146,23 @@ before_install:
...
-  - travis_retry pip install gmpy2  # speeds up mpmath (scipy.special tests)
-  - travis_retry pip install pybind11
   - |
     if [ -z "${USE_DEBUG}" -a "${TRAVIS_CPU_ARCH}" == "amd64" -a "${TRAVIS_PYTHON_VERSION}" != "3.8" ]; then
         travis_retry pip install numba==0.47
         travis_retry pip install --only-binary=:all: sparse
     fi
   - |
-    if [ "${TESTMODE}" == "full" ]; then
+    if [ "${TESTMODE}" == "full"  ]; then
```
Comment (rgommers) [inline .travis.yml]: reverted this change as well to minimize the diff

File: .travis.yml [hunk 15]
```diff
@@ -199,8 +226,14 @@ script:
         cd ..
         USE_WHEEL_BUILD="--no-build"
     fi
-  - export SCIPY_AVAILABLE_MEM=3G
-  - python -u runtests.py -g -m $TESTMODE $COVERAGE $USE_WHEEL_BUILD -- -rfEX -n 3 2>&1 | tee runtests.log
+  - |
+    if [[ "$TRAVIS_CPU_ARCH" == "arm64" ]]; then
+      python -u runtests.py --parallel=4 -g -m $TESTMODE $COVERAGE $USE_WHEEL_BUILD -- -rfEX 2>&1 | tee runtests.log
+    else
+      export SCIPY_AVAILABLE_MEM=3G
```
Comment (rgommers) [inline .travis.yml:233]: this should also apply to the arm64 test

**Comment by rgommers** [comment]
Timed out at 60% this time. It looks like this really won't work reliably. One thing that could make sense is to only run CI for some submodules - that way we still test the build and some of the modules with the most fragile compiled code. I'm thinking `special`, `ndimage`, `sparse`, `fft` for starters. And for PRs that touch code we may want to test in other modules, we then have an easy way to switch things around and run ARM64 CI for just that module.

Thoughts?

**Comment by odidev** [comment]
@rgommers The build is still not running successfully, any suggestions on what could be done to make it work. Also can we remove/skip some part of code for arm64 so that build time could decrease?

**Comment by rgommers** [comment]
> any suggestions on what could be done to make it work.

see my suggestion above on running only part of the test suite, only for some modules like `special` etc. 

> Also can we remove/skip some part of code for arm64 so that build time could decrease?

We could but I'd rather not if we can avoid it. Skipping `sparse` would save a lot of time in particular, but a full build + partial test suite run is preferable to a partial build.

**Comment by andyfaff** [comment]
The current setup in the PR is to `USE_SDIST=1`. An excerpt from the travis.yml file says:
```
 if [ "${USE_WHEEL}" == "1" ]; then
        # Run setup.py build before pip wheel, to build in current directory
        # and make more efficient use of ccache
        echo "setup.py build"
        python tools/suppress_output.py python setup.py build
        echo "pip wheel"
        python tools/suppress_output.py pip wheel --no-build-isolation .
        pip install --no-cache-dir scipy*.whl
        USE_WHEEL_BUILD="--no-build"
    elif [ "${USE_SDIST}" == "1" ]; then
        echo "setup.py sdist"
        python tools/suppress_output.py python setup.py sdist
        # Move out of source directory to avoid finding local scipy
```
From the travis logs the build does use sdist. The comments indicating using USE_WHEEL would make the caching better? (faster build and test).

I wonder if a full-build and randomly chosen submodule for testing would be acceptable?

**Comment by andyfaff** [comment]
Another option could be to cross-compile on a Linux instance using the `crossbuild-essential-arm64` (it might be faster), and then to test that artifact in a second step on an arm64 box.

For example, circleCI builds the docs then in another step deploys the documents.
I wonder if that kind of two step process is possible on travis.

One thing is for sure, the travis.yml file could probably do with a bit of cleaning up.

**Comment by andyfaff** [comment]
travisCI does [have build stages](https://docs.travis-ci.com/user/build-stages/). If we reorganise the file somehow we could build all the artifacts in step 1, then test them in step 2. I'm guessing that the timeouts would be alleviated.

**Comment by rgommers** [comment]
> From the travis logs the build does use sdist. The comments indicating using USE_WHEEL would make the caching better? (faster build and test).

Nice idea. I'm not sure that improvement will always materialize or last forever (since the `pip` devs plan to make wheel builds always go via an sdist in the future). But it is worth doing anyway, it at least can make timeouts less likely.

> Another option could be to cross-compile on a Linux instance using the `crossbuild-essential-arm64` (it might be faster), and then to test that artifact in a second step on an arm64 box.

`distutils` has very little support for cross-compiling, and the people that have been trying, e.g. for Raspbian, usually run into issue. I wouldn't want to rely on that.


> travisCI does [have build stages](https://docs.travis-ci.com/user/build-stages/). If we reorganise the file somehow we could build all the artifacts in step 1, then test them in step 2. I'm guessing that the timeouts would be alleviated.

That would have multiple benefits other than ARM64 support. But is also pretty significant surgery - probably best to not mix the two. 

Going with `USE_WHEEL` + limit to a few modules, and then switching to build stages when someone wants to work on that, seems like the way to go here.

**Comment by andyfaff** [comment]
I'm not sure if the time for the individual build stages are combined for a specific matrix entry or not.

**Comment by andyfaff** [comment]
If I look at the last build log I'm pretty sure that numpy is being built as well, see [L5005](https://travis-ci.org/github/scipy/scipy/jobs/680533100#L5005). That's going to add a long time to the build.

So, the prebuilt numpy package isn't usable for some reason. For macOS/windows we currently rely on downloading openblas. I'm looking into that now. 
Another route could be to build a suitable numpy wheel and stash it somewhere. Alternatively, it's looking like the next numpy release will have arm64 wheels, so we could wait until then.

**Comment by andyfaff** [comment]
Ok, so I've got it running successfully now.

**Comment by andyfaff** [comment]
@odidev, @rgommers, I made some changes that work on my fork. As part of those changes I had to rebase on master because there was a merge conflict. When I pushed my changes to odidev's branch I force pushed them. The commits have the right authorship, as viewed from gitk on my computer and under the commit tab above, but in the commit tab above they say they were committed by me. Hopefully I haven't screwed any attribution up, that wasn't my intention.

**Comment by andyfaff** [comment]
It's not going to close the original issue. That issue refers to wheels being produced for distribution. This PR solely tests arm64 for CI purposes.

**Comment by andyfaff** [comment]
For some reason travisCI isn't triggering, but it is on my fork. https://travis-ci.org/github/andyfaff/scipy/builds/681422395?utm_source=github_status&utm_medium=notification

**Comment by andyfaff** [comment]
Close/reopen trick worked.

**Comment by andyfaff** [comment]
> Alternatively, it's looking like the next numpy release will have arm64 wheels, so we could wait until then.

I may be wrong about that, see https://github.com/MacPython/numpy-wheels/pull/74#issuecomment-621753471

**Comment by rgommers** [comment]
> It's not going to close the original issue. That issue refers to wheels being produced for distribution. This PR solely tests arm64 for CI purposes.

Yep, that's as far as we can go, and is a big step forward. If people need ARM64 binaries, they should use https://github.com/conda-forge/miniforge

File: .travis.yml [hunk 16]
```diff
@@ -201,8 +206,10 @@ before_install:
       conda install pip conda;
       conda update -n base conda;
       conda info -a;
-      conda install pytest pytest-xdist $NUMPYSPEC mpmath Cython pybind11;
+      conda install pytest pytest-xdist numpy mpmath Cython pybind11;
```
Comment (rgommers) [inline .travis.yml:209]: While unpinning to use the latest numpy release sounds fine, I think this should remain as `$NUMPYSPEC`.

**Comment by rgommers** [comment]
> If I look at the last build log I'm pretty sure that numpy is being built as well, see L5005. That's going to add a long time to the build.

Good catch, we should grab a NumPy wheel from the new Anaconda storage.

> So, the prebuilt numpy package isn't usable for some reason. For macOS/windows we currently rely on downloading openblas. 

We need to build against OpenBLAS, can't just rely on what's bundled in with NumPy. There's nothing wrong here I think.

**Comment by andyfaff** [comment]
> We need to build against OpenBLAS, can't just rely on what's bundled in with NumPy. There's nothing wrong here I think.

sorry, I meant to say that for macOS we grab a stashed library build, we could do a similar thing for a numpy wheel. In any case it should be working now.

File: .travis.yml [hunk 17]
```diff
@@ -145,18 +161,18 @@ before_install:
     fi
   - python -V -V
   - python -c 'import sys; print("Python debug build:", hasattr(sys, "gettotalrefcount"))'
-  - travis_retry pip install --upgrade pip setuptools wheel
-  - travis_retry pip install cython
-  - if [ -n "$NUMPYSPEC" ]; then travis_retry pip install $NUMPYSPEC; fi
-  - travis_retry pip install --upgrade pytest pytest-xdist mpmath argparse Pillow codecov
+  # arm64 build uses conda (there aren't any arm64 wheels on PyPI, need
```
Comment (hrw) [inline .travis.yml:164]: Run a build where you build those wheels. Add pip cache into cache (it is even there already).

Do another run - pip should take those wheels from local cache.

Should work iirc.

File: .travis.yml [hunk 18]
```diff
@@ -209,12 +241,20 @@ script:
         USE_WHEEL_BUILD="--no-build"
     fi
   - export SCIPY_AVAILABLE_MEM=3G
-  - python -u runtests.py -g -m $TESTMODE $COVERAGE $USE_WHEEL_BUILD -- -rfEX -n 3 2>&1 | tee runtests.log
+  - |
+    if [[ "$TRAVIS_CPU_ARCH" == "arm64" ]]; then
```
Comment (hrw) [inline .travis.yml:245]: Why not running 4 threads on all architectures?

Comment (pv) [reply .travis.yml:245]: Does Travis have 4 cores on all archs?

Comment (andyfaff) [reply .travis.yml:245]: travisCI has 2 cores on all archs. https://docs.travis-ci.com/user/reference/overview/

**Comment by rgommers** [comment]
Ah I forgot, `--parallel` gets disabled in `setup.py`:
```
            # Disable distutils parallel build, due to race conditions
            # in numpy.distutils (Numpy issue gh-15957)
            if self.parallel:
                print("NOTE: -j build option not supported. Set NPY_NUM_BUILD_JOBS=4 "
                      "for parallel build.")
            self.parallel = None
```

So building is non-parallel, and hence `--parallel` argument to `runtests.py` is only used for testing. Which is equally well done with `-- -n` - so let's use that also for arm64.

File: .travis.yml (hunk 18 above)
Comment (rgommers) [reply .travis.yml:245]: We now use `-n 3`. I assume we tested that that was faster than `-n 2` at some point. We should probably do the same here for arm64 as for x86, no need to deviate.

**Comment by hrw** [comment]
```
E   Max absolute difference: 1.80396576e-07
E   Max relative difference: 9.05137079e-08
```

Looks like VFP precision difference? 

I had that few years ago with some software: https://marcin.juszkiewicz.com.pl/2015/09/19/from-a-diary-of-aarch64-porter-vfp-precision/

**Comment by hrw** [comment]
Have you considered caching "https://github.com/conda-forge/miniforge/releases/download/4.8.2-1/Miniforge3-4.8.2-1-Linux-aarch64.sh" and following Conda downloads?

**Comment by andyfaff** [comment]
There don't seem to be timeout issues now, it's finishing well within time. If the entry starts getting flaky then it might be worth doing some work on this, but otherwise I think it's ok

**Comment by rgommers** [comment]
On the last run it didn't start, errored after 48 sec. Restarting now.

File: .travis.yml (hunk 18 above)
Comment (rgommers) [reply .travis.yml:245]: This comment isn't addressed yet - the whole `if-else` seems unnecessary. This `runtests.py` command doesn't need touching at all.

Comment (rgommers) [reply .travis.yml]: pushed a change

**Comment by rgommers** [comment]
> Looks like VFP precision difference?

It's a new test, and the test precision is simply a tiny bit too tight it looks like. I pushed a commit that should fix it.

**Comment by rgommers** [comment]
The last run with `--parallel=4` took 49 minutes, let's see if it's similar with `-n 3`.

**Comment by rgommers** [comment]
Passed in 24 minutes - guess traffic is lighter on the weekend:) Just something wrong with the `mpmath` check.

**Comment by rgommers** [comment]
Hmm that failed on `apt-get` not finding `libmpc-dev`, looks like a flake. Re-running again.

**Review by rgommers — APPROVED** [review]

**Comment by rgommers** [comment]
33 min, not much slower than the AMD64 jobs:

image

All green, so let's call this good and merge it.

**Comment by rgommers** [comment]
Thanks so much @odidev for working on this! And @andyfaff and everyone else for pitching in!

**Comment by andyfaff** [comment]
Using miniconda is much slower than pip installing into a system python but it looks ok for now.

**Comment by rgommers** [comment]
> Using miniconda is much slower than pip installing into a system python but it looks ok for now.

On the last CI run the whole Miniforge + all dependencies install took 94 sec., the build took 382 sec. and running the test suite 1070 sec., so I don't think conda vs pip matters much.

**Comment by mdhaber** [comment]
> see my suggestion above on running only part of the test suite
@rgommers is this running the whole test suite after all?

**Comment by rgommers** [comment]
> @rgommers is this running the whole test suite after all?

`scipy.test()` not `scipy.test('full')`

**Comment by rgommers** [comment]
but yes, all modules

**Comment by odidev** [comment]
> Thanks so much @odidev for working on this! And @andyfaff and everyone else for pitching in!

Thanks for accepting this PR @rgommers!

**Comment by rgommers** [comment]
This job is unfortunately timing out regularly. Multiple tests over the weekend looked good, but the timing is hugely variable. When timeouts happen the test suite is almost through (99% on the last PR I checked), but still it's annoying. We'll have to do something about it. Either speeding up the whole test suite again by hunting down the slowest tests (gh-12014 will help with that), or disabling some tests/modules on `arm64`.

**Comment by andyfaff** [comment]
I think it's the build that taking a long time.  It's on the order of >30mins for the times it fails. Setup is pretty fast.

**Comment by rgommers** [comment]
You're right, the variation there is 5 min - 30 min. Let's see if gh-12026 helps.

**Comment by tylerjereddy** [comment]
Oh, this is already in the `1.5.x` branch.. so ARM64 CI is passing on that branch with Python 3.7 then I guess...
